### ISSUE TITLE
GumbelVQ 1 Epoch on Open Images Localized Annotations

### ISSUE THREAD

**Issue by afiaka87** [robvanvolt/DALLE-models#7]
edit: Well this thing trained for a full 5 epochs and never made a single coherent generation. @rom1504 and I discussed this and I guess it's just really really hard to train this one. 🤷  Don't think a single RTX 2070 is going to cut it.

Hey - I'm trying to force myself to take a few days off and the holidays are coming up here in the states anyways so I'll probably be away from the discord/github for a bit.

## use right away with python
```sh
pip install wandb requests
```
```python
import wandb, requests
run = wandb.init()
artifact = run.use_artifact('dalle-pytorch-replicate/oi_gumbel_imgloss7/trained-dalle:v0', type='model')
artifact_dir = artifact.download()

import zipfile
with zipfile.ZipFile(path_to_zip_file, 'r') as zip_ref:
    zip_ref.extractall(directory_to_extract_to)
openaiblog_openimages_bpe_url = "https://github.com/robvanvolt/DALLE-models/files/6735615/blogoimixer_4096.bpe.zip"
downloaded_obj = requests.get(openaiblog_openimages_bpe_url)
with open(openaiblog_openimages_bpe_url, "wb") as file:
      file.write(downloaded_obj.content)
```

## Download Links
- https://github.com/robvanvolt/DALLE-models/files/6735615/blogoimixer_4096.bpe.zip
- https://api.wandb.ai/artifactsV2/gcp-us/dalle-pytorch-replicate/QXJ0aWZhY3Q6MTI4MTE0MzU=/48aa0681b9ee25dc6711b234454d364d

## Info
- data - [Open Images Localized Annotations](https://www.dropbox.com/s/8rue7r9ppds3jyk/open_images_localized_annotations.tar.gz?dl=0)
- variational autoencoder -  Gumbel (f=8) VQGAN
- tokenizer 
  - [blogoimixer_4096.bpe.zip](https://github.com/robvanvolt/DALLE-models/files/6735615/blogoimixer_4096.bpe.zip)

```sh
pip3 install youtokentome # will install the cli tool `yttm`
yttm bpe --vocab_size=4096 --coverage=1.0 --model=blogoimixer_4096.bpe --data=blogoi_allcaps.txt
```
- optimizer
`deepspeed.runtime.ops.Adam(3e-4, betas=(0.9, 0.96), eps=1e-8))`
- scheduler
```python
# Warmup the learning rate from 1e-6 to 4e-3 for 2% of all steps then 
# do Cosine decay back down to 1e-6 and train there until finished.
# Requires knowledge of the global/total step count. So you need to calculate it. 
# Also requires deepspeed if you aren't into that.
total_num_steps = int(len(ds) * EPOCHS / (BATCH_SIZE * args.ga_steps)) # number of image-text pairs in a single epoch multiplied by the effective batch size
warmup_num_steps = total_num_steps * 0.02
deepspeed.runtime.ops.schedulers.WarmupLRDecay(
    total_num_steps=total_num_steps,
    warmup_steps=warmup_num_steps,
    # CogView did 0 here? Anyway I stuck with @janEbert 1e-6 because a 0 learning rate doesn't improve anything. 
    min_lr=1e-6, 
    max_lr=LEARNING_RATE,
)
```
- clip_grad_norm 1.0
- learning_rate 3e-4
- batch_size 32
- truncate_captions
- random_resize_crop_lower_ratio 1.0
- dim 512 
- text_seq_len 256 
- depth 12 
- heads 12
- dim_head 64
- ff_dropout 0.1
-  attn_dropout 0.1
- attn_types full,axial_row,full,axial_col,axial_row,axial_row,axial_col,axial_row,axial_row,axial_row,axial_col,conv_like

**Comment by afiaka87** [comment]
Anyway I'm still training and I'll be away for awhile - but I'll submit a proper PR once I have a resumable DeepSpeed checkpoint ready for upload. in the meantime you can direct people to the public W&B training session. Seems like one epoch takes about a day. I'll try to finish another one but who knows. 

https://wandb.ai/dalle-pytorch-replicate/oi_gumbel_imgloss7/runs/ml9u03ab

**Comment by afiaka87** [comment]
also probably best to keep in mind that the captions in this one are spoken word transcribe. Folks from CLIP-land might have to readjust their expectations. Could be a good idea to provide sort of "caption template".

"In this image I can see <blank> and over here <blank> and there are <blank> and to the left <blank> and to the right <blank> and at the top <blank> and to the bottom <blank>" 

where possible fillings would be


"In this image I can see cars and over here a fence and there are signs and to the left buildings and to the right a field and at the top the night sky and to the bottom a sidewalk"

I think i'm most interested in seeing how well it manages to learn to compose "left, right, top, and bottom" into the correct positions while maintaining visual fidelity. Should be interesting.

**Comment by robvanvolt** [comment]
The loss!! *_* *jealous*

**Comment by afiaka87** [comment]
Oh just to be clear that loss graph is from a session where I used `--loss_img_weight 1`, and this one used the default value of 7.

Since we aren't logging both the text-predicts-text loss and the text-predicts-image loss - it's not really a meaningful comparison because the text-predicts-text loss presumably overfits and goes down _very_ quickly unless your captions dataset is as diverse as say, CogView.

**Comment by afiaka87** [comment]
> The loss!! _jealous_

![W B Chart 6_29_2021, 2 23 29 PM](https://user-images.githubusercontent.com/3994972/123855589-b1618200-d8e5-11eb-9e7d-fd3a52a57393.png)

**Comment by afiaka87** [comment]
## Details:

I found in another experiment that using a little bit of full attention acutally helps a good deal with the GumbelVQ - 

Here;  I think this shows it. Two runs. Exact same parameters except for this:
```
desert-plasma-29
full,axial_row,axial_col,conv_like,full,axial_row,axial_col,conv_like
depth 8

logical-jazz-19
depth 12 
axial_row,axial_row,axial_col,axial_row,axial_row,axial_row,axial_col,axial_row,axial_row,axial_row,axial_col,conv_like
```

Per usual, using axial attention increase my `samples per second` from `11` to `22` - so the x-axis here is time instead of step.

![W B Chart 6_29_2021, 1 09 25 PM](https://user-images.githubusercontent.com/3994972/123846640-401cd180-d8db-11eb-92fe-299ef876a930.png)

As you can see - lower depth of 8 with 2 layers of full attention outperforms 12 layers of sparse axial attention.

**Comment by afiaka87** [comment]
@robvanvolt probably worth mentioning (since i havent seen generations from it yet) that it's perhaps not super likely to work great. this thing might need a bigger DALL-E than I can train. I never got to anything resembling the caption in my fp32 training once. And it was very often just gumbelvq non-sense output. It seems to be a very large codebook so the transformer needs a lot more time to find all the codes in it.

**Comment by afiaka87** [comment]
edit: Well this thing trained for a full 5 epochs and never made a single coherent generation. @rom1504 and I discussed this and I guess it's just really really hard to train this one. 🤷 Don't think a single RTX 2070 is going to cut it.
